openjdk/mmtkVMCompanionThread.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 1998, 2017, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "mmtk.h"
+#include "mmtkVMCompanionThread.hpp"
+#include "runtime/mutex.hpp"
+
+MMTkVMCompanionThread::MMTkVMCompanionThread():
+    NamedThread(),
+    _desired_state(_threads_resumed),
+    _reached_state(_threads_resumed),
+    _resumption_count(0) {
+  set_name("MMTK VM Companion Thread");
+  _lock = new Monitor(Monitor::nonleaf,
+                      "MMTkVMCompanionThread::_lock",
+                      true,
+                      Monitor::_safepoint_check_never);
+}
+
+MMTkVMCompanionThread::~MMTkVMCompanionThread() {
+  guarantee(false, "MMTkVMCompanionThread deletion must fix the race with VM termination");
+}
+
+void MMTkVMCompanionThread::run() {
+  this->initialize_named_thread();
+
+  for (;;) {
+    // Wait for suspend request
+    log_trace(gc)("MMTkVMCompanionThread: Waiting for suspend request...");
+    {
+      MutexLockerEx locker(_lock, Mutex::_no_safepoint_check_flag);
+      assert(_reached_state == _threads_resumed, "Threads should be running at this moment.");
+      while (_desired_state != _threads_suspended) {
+        _lock->wait(true);
+      }
+      assert(_reached_state == _threads_resumed, "Threads should still be running at this moment.");
+    }
+
+    // Let the VM thread stop the world.
+    log_trace(gc)("MMTkVMCompanionThread: Letting VMThread execute VM op...");
+    VM_MMTkSTWOperation op(this);
+    VMThread::execute(&op);
```

Comment:
Is `execute()` blocking? We probably want to have some comments here.
```suggestion
    // execute() is blocking. The companion thread will be blocked here and the VM thread will be blocked in reach_suspended_and_wait_for_resume(), until a request() for resume comes in.
    VMThread::execute(&op);
```

Reply:
For JDK 11, not always.  It depends on the `evaluation_mode()` and `evaluation_concurrently()`.  There is only one non-blocking VMOperation in JDK 11.  The `VM_MMTkSTWOperation` operation I defined is blocking.

For JDK 17, all VM operations are blocking.

Anyway I think it is helpful to add a comment.